intake/source/csv.py

```diff
@@ -81,46 +84,115 @@ def _path_column(self):
             return
 
         drop_path_column = "include_path_column" not in self._csv_kwargs
         path_column = self._path_column()
 
-        self._dataframe = dask.dataframe.read_csv(urlpath, storage_options=self._storage_options, **self._csv_kwargs)
+        self._dask_df = dask.dataframe.read_csv(urlpath, storage_options=self._storage_options, **self._csv_kwargs)
 
         # add the new columns to the dataframe
-        self._set_pattern_columns(path_column)
+        self._dask_df = self._set_pattern_columns(self._dask_df, path_column)
 
         if drop_path_column:
-            self._dataframe = self._dataframe.drop([path_column], axis=1)
+            self._dask_df = self._dask_df.drop([path_column], axis=1)
+
+    def files(self):
+        if self._files is not None:
+            return self._files
 
-    def _get_schema(self):
         urlpath = self._get_cache(self._urlpath)[0]
+        if not isinstance(urlpath, list):
+            urlpath = [urlpath]
 
-        if self._dataframe is None:
-            self._open_dataset(urlpath)
+        glob_in_path = any("*" in path for path in urlpath)
+        if self.pattern is None and not glob_in_path:
+            self._files = urlpath
+        else:
+            from fsspec.core import get_fs_token_paths
 
-        dtypes = self._dataframe._meta.dtypes.to_dict()
-        dtypes = {n: str(t) for (n, t) in dtypes.items()}
-        return base.Schema(dtype=dtypes, shape=(None, len(dtypes)), npartitions=self._dataframe.npartitions, extra_metadata={})
+            self._files = get_fs_token_paths(urlpath)[2]
+
+        return sorted(self._files)
+
+    def _get_schema(self):
+        if self._schema is not None:
+            return self._schema
+
+        if self._dask_df is not None:
+            dtypes = self._dask_df._meta.dtypes.to_dict()
+            dtypes = {n: str(t) for (n, t) in dtypes.items()}
+            self._schema = base.Schema(dtype=dtypes, shape=(None, len(dtypes)), npartitions=self._dask_df.npartitions, extra_metadata={})
+            return self._schema
+
+        nrows = self._csv_kwargs.get("nrows")
+        self._csv_kwargs["nrows"] = 10
+        df = self._get_partition(0)
+        if nrows is None:
+            del self._csv_kwargs["nrows"]
+        else:
+            self._csv_kwargs["nrows"] = nrows
+
+        dtypes = {col: str(dtype) for col, dtype in df.dtypes.items()}
+        self._schema = base.Schema(dtype=dtypes, shape=(None, len(dtypes)), npartitions=len(self.files()), extra_metadata={})
+
+        return self._schema
 
     def _get_partition(self, i):
-        self._get_schema()
-        return self._dataframe.get_partition(i).compute()
+        if self._dask_df is not None:
+            # FIX: side effects from having different partition
+            # definitions (dask vs. files)?
+            return self._dask_df.get_partition(i).compute()
+
+        url_part = self.files()[i]
```

Comment:
+1

Reply:
as in `url_part = self.files()[i+1]`? Dask dataframe [is 0-indexed](https://docs.dask.org/en/stable/generated/dask.dataframe.DataFrame.get_partition.html)...

Reply:
No, "+1" was meant to express agreement (like 👍 ), since this is a specific thing we talked about.